# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [1]:
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
import numpy as np
from sklearn.preprocessing import PolynomialFeatures
tips = sns.load_dataset('tips')
tips['tip_pct'] = tips['tip'] / tips['total_bill']
net_bill = tips['total_bill'] - tips['tip']
tip_pct = tips['tip_pct']
net_bill = sm.add_constant(net_bill)
modelo = sm.OLS(tip_pct, net_bill).fit()
modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.189
Model:                            OLS   Adj. R-squared:                  0.185
Method:                 Least Squares   F-statistic:                     56.25
Date:                Tue, 12 Mar 2024   Prob (F-statistic):           1.21e-12
Time:                        13:58:45   Log-Likelihood:                 361.93
No. Observations:                 244   AIC:                            -719.9
Df Residuals:                     242   BIC:                            -712.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2162      0.008     26.401      0.000       0.200       0.232
0             -0.0033      0.000     -7.500      0.000      -0.004      -0.002
==============================================================================
Omnibus:                      227.090   Durbin-Watson:                   1.990
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8183.493
Skew:                           3.464   Prob(JB):                         0.00
Kurtosis:                      30.512   Cond. No.                         43.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [2]:
# Aplicando o logaritmo à variável net_bill
log_net_bill = np.log(tips['total_bill'] - tips['tip'])
# Adicionando uma constante à variável independente para a regressão
log_net_bill = sm.add_constant(log_net_bill)
# Realizando a regressão de tip_pct explicada pelo logaritmo de net_bill
modelo_log = sm.OLS(tip_pct, log_net_bill).fit()
modelo_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.289
Model:                            OLS   Adj. R-squared:                  0.286
Method:                 Least Squares   F-statistic:                     98.21
Date:                Tue, 12 Mar 2024   Prob (F-statistic):           1.19e-19
Time:                        13:58:45   Log-Likelihood:                 377.99
No. Observations:                 244   AIC:                            -752.0
Df Residuals:                     242   BIC:                            -745.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.3468      0.019     18.198      0.000       0.309       0.384
0             -0.0686      0.007     -9.910      0.000      -0.082      -0.055
==============================================================================
Omnibus:                      169.023   Durbin-Watson:                   2.010
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2946.559
Skew:                           2.432   Prob(JB):                         0.00
Kurtosis:                      19.314   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [3]:
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(tips['total_bill'].values.reshape(-1, 1))
# Realizando a regressão de tip_pct explicada pelo polinômio de segundo grau em net_bill
modelo_poly = sm.OLS(tip_pct, X_poly).fit()
modelo_poly.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.140
Model:                            OLS   Adj. R-squared:                  0.133
Method:                 Least Squares   F-statistic:                     19.65
Date:                Tue, 12 Mar 2024   Prob (F-statistic):           1.24e-08
Time:                        13:58:45   Log-Likelihood:                 354.86
No. Observations:                 244   AIC:                            -703.7
Df Residuals:                     241   BIC:                            -693.2
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2525      0.019     13.113      0.000       0.215       0.290
x1            -0.0068      0.002     -3.957      0.000      -0.010      -0.003
x2          8.951e-05   3.35e-05      2.676      0.008    2.36e-05       0.000
==============================================================================
Omnibus:                      213.171   Durbin-Watson:                   2.072
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6542.764
Skew:                           3.192   Prob(JB):                         0.00
Kurtosis:                      27.552   Cond. No.                     3.47e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.47e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [4]:
# Aplicando o logaritmo ao tip_pct
log_tip_pct = np.log(tip_pct)
# Adicionando uma constante à variável independente para a regressão
net_bill_with_const = sm.add_constant(net_bill)
# Realizando a regressão do logaritmo do tip_pct explicado por net_bill
modelo_log_tip_pct = sm.OLS(log_tip_pct, net_bill_with_const).fit()
# Calculando o R² para o modelo do logaritmo do tip_pct
R_squared_log_tip_pct = modelo_log_tip_pct.rsquared
# Calculando o R² para o modelo do tip_pct original
R_squared_tip_pct = 1 - (modelo_log_tip_pct.resid.var() / tip_pct.var())
print("R² para o modelo do logaritmo do tip_pct:", R_squared_log_tip_pct)
print("R² para o modelo do tip_pct original:", R_squared_tip_pct)

R² para o modelo do logaritmo do tip_pct: 0.2436292435963786
R² para o modelo do tip_pct original: -23.878538834375476


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [5]:
log_net_bill = np.log(tips['total_bill'] - tips['tip'] + 1e-6)
# Aplicando o logaritmo ao tip_pct
log_tip_pct = np.log(tip_pct)
# Adicionando uma constante à variável independente para a regressão
log_net_bill_with_const = sm.add_constant(log_net_bill)
# Realizando a regressão do logaritmo do tip_pct explicado pelo logaritmo de net_bill
modelo_log_log = sm.OLS(log_tip_pct, log_net_bill_with_const).fit()
# Calculando o R² para o modelo do logaritmo do tip_pct
R_squared_log_log = modelo_log_log.rsquared
# Calculando o R² para o modelo do tip_pct original
residuals_tip_pct = np.exp(log_tip_pct) - np.exp(modelo_log_log.predict(log_net_bill_with_const))
R_squared_tip_pct = 1 - (residuals_tip_pct.var() / tip_pct.var())
print("R² para o modelo do logaritmo do tip_pct:", R_squared_log_log)
print("R² para o modelo do tip_pct original:", R_squared_tip_pct)

R² para o modelo do logaritmo do tip_pct: 0.2799741825326272
R² para o modelo do tip_pct original: 0.3340210015872478


###  6. qual modelo resultou no melhor ajuste?

Comparando os valores de $R^2$ observamos que o modelo 4, ajustado com tip_pct em função do logaritmo de net_bill, apresentou o melhor ajuste com um $R^2$ de 0.2890. é o que resultou no melhor ajuste entre os modelos analisados.